crates/oxc_linter/src/rules/unicorn/consistent_function_scoping.rs

```diff
@@ -0,0 +1,1233 @@
+                        Span::new(function.span.start, function.span.start + 8),
+                    )
+                } else if let Some(function_id) = &function.id {
+                    (function_id.symbol_id.get().unwrap(), function_body, function_id.span())
+                } else {
+                    return;
+                }
+            }
+            AstKind::ArrowFunctionExpression(arrow_function) => {
+                if !self.check_arrow_functions {
+                    return;
+                }
+                if let Some(binding_ident) = get_function_like_declaration(node, ctx) {
+                    (
+                        binding_ident.symbol_id.get().unwrap(),
+                        &arrow_function.body,
+                        binding_ident.span(),
+                    )
+                } else {
+                    return;
+                }
+            }
+            _ => return,
+        };
+        // if the function is declared at the root scope, we don't need to check anything
+        if ctx.symbols().get_scope_id(function_declaration_symbol_id)
+            == ctx.scopes().root_scope_id()
+        {
+            return;
+        }
+
+        if let Some(AstKind::ReturnStatement(_)) =
+            outermost_paren_parent(node, ctx).map(oxc_semantic::AstNode::kind)
+        {
+            return;
+        }
+
+        if is_parent_scope_iife(node, ctx) || is_in_react_hook(node, ctx) {
+            return;
+        }
+
+        // get all references in the function body
+        let (function_body_var_references, is_parent_this_referenced) = {
+            let mut rf = ReferencesFinder::default();
+            rf.visit_function_body(function_body);
+            (rf.references, rf.is_parent_this_referenced)
+        };
+
+        if is_parent_this_referenced && matches!(node.kind(), AstKind::ArrowFunctionExpression(_)) {
+            return;
+        }
+
+        let parent_scope_ids = {
+            let mut current_scope_id = ctx.symbols().get_scope_id(function_declaration_symbol_id);
+            let mut parent_scope_ids = FxHashSet::default();
+            parent_scope_ids.insert(current_scope_id);
+            while let Some(parent_scope_id) = ctx.scopes().get_parent_id(current_scope_id) {
+                parent_scope_ids.insert(parent_scope_id);
+                current_scope_id = parent_scope_id;
+            }
+            parent_scope_ids
+        };
+
+        for reference_id in function_body_var_references {
+            let reference = ctx.symbols().get_reference(reference_id);
+            let Some(symbol_id) = reference.symbol_id() else { continue };
+            let scope_id = ctx.symbols().get_scope_id(symbol_id);
+            if parent_scope_ids.contains(&scope_id) && symbol_id != function_declaration_symbol_id {
+                return;
+            }
+        }
+
+        ctx.diagnostic(
+            OxcDiagnostic::warn("Function does not capture any variables from the outer scope.")
```

Comment:
Diagnostic creation should be in a separate function.

Reply:
fixed